packages/flutter/lib/src/rendering/selection.dart

```diff
@@ -100,17 +100,93 @@ abstract class SelectionHandler implements ValueListenable<SelectionGeometry> {
   SelectionResult dispatchSelectionEvent(SelectionEvent event);
 }
 
+/// A controller for the content of a [Selectable].
+///
+/// A user can set the value on this controller to modify the content under
+/// a selection created by [SelectionArea] or [SelectableRegion].
+abstract class SelectedContentController<T extends Object> extends ValueNotifier<T> {
+  /// Creates a controller for the content of a [Selectable] or [SelectionHandler].
+  SelectedContentController(super.value);
+
+  /// Additional controllers to include as children.
+  ///
+  /// Children of a given controller enable more granular modification of the
+  /// selection.
+  List<SelectedContentController<Object>> children = <SelectedContentController<Object>>[];
+
+  /// The start of the selection relative to the type of content in [value].
+  int get startOffset;
+
+  /// The end of the selection relative to the type of content in [value].
+  int get endOffset;
+
+  /// Builds the content from current [value].
+  T buildContents();
+
+  /// Adds a child controller to the list of [children].
+  void addChild(SelectedContentController<Object> childController) {
+    children.add(childController);
+  }
+
+  @override
+  String toString() {
+    return 'SelectedContentController(\n'
+           '  value: $value,\n'
+           '  startOffset: $startOffset,\n'
+           '  endOffset: $endOffset,\n'
+           '  children: $children,\n'
+           ')';
+  }
+}
+
 /// The selected content in a [Selectable] or [SelectionHandler].
 // TODO(chunhtai): Add more support for rich content.
 // https://github.com/flutter/flutter/issues/104206.
 class SelectedContent {
   /// Creates a selected content object.
   ///
   /// Only supports plain text.
-  const SelectedContent({required this.plainText});
+  const SelectedContent({
+    required this.plainText,
+    required this.geometry,
+    this.startOffset = -1,
+    this.endOffset = -1,
+    this.controllers,
+  });
 
   /// The selected content in plain text format.
   final String plainText;
+
+  /// The [SelectionGeometry] of the [Selectable] or [SelectionHandler] containing
+  /// the selection.
+  final SelectionGeometry geometry;
+
+  /// The value representing the beginning of the selection, defaults to -1.
+  ///
+  /// If the [Selectable] contains mutable text, then the offset represents
```

Comment:
I can do this, I assumed the `SelectionDelegate` provides the selectable index through `currentSelectionStartIndex` and `currentSelectionEndIndex`.

Reply:
I figure using a flatten in may be harder for calculation because now you kinda have to add the length of selectable together? Also for developer that uses this api they also need to enumerate all content in the selectables to be able to find the right spot

Reply:
I'm not sure I understand. What calculation would be harder without providing the selectable index, by index do you mean the index of a selectable within a given `SelectionDelegate`?

> Also for developer that uses this api they also need to enumerate all content in the selectables to be able to find the right spot
We only return the `SelectedContentController` for selectables with an active selection, so they would only need to iterate over those and find a spot based on that.

Reply:
These offsets have now been removed from `SelectedContent`.